### PR TITLE
bpo-34088: sndhdr.what() doesn't fail on bad input. 

### PR DESCRIPTION
**Pull request by davidawad**
## What this is 
This is a very small fix for `sndhdr.py` so it can take bad input and not croak. This was originally reported by Jussi Judin on BPO [here](https://bugs.python.org/issue34088).

To reproduce this these crashes you can try the following examples:

```
$ echo UklGRjAwMDBXQVZFZm10IDAwMDABADAwMDAwMDAwMDAwMDAw | python3.7 -mbase64 -d > in.file
$ python3.7 sndhdr/test.py in.file
Traceback (most recent call last):
  File "sndhdr/test.py", line 4, in <module>
    sndhdr.what(sys.argv[1])
  File "/tmp/python-3.7-bin/lib/python3.7/sndhdr.py", line 54, in what
    res = whathdr(filename)
  File "/tmp/python-3.7-bin/lib/python3.7/sndhdr.py", line 63, in whathdr
    res = tf(h, f)
  File "/tmp/python-3.7-bin/lib/python3.7/sndhdr.py", line 163, in test_wav
    w = wave.open(f, 'r')
  File "/tmp/python-3.7-bin/lib/python3.7/wave.py", line 510, in open
    return Wave_read(f)
  File "/tmp/python-3.7-bin/lib/python3.7/wave.py", line 164, in __init__
    self.initfp(f)
  File "/tmp/python-3.7-bin/lib/python3.7/wave.py", line 153, in initfp
    chunk.skip()
  File "/tmp/python-3.7-bin/lib/python3.7/chunk.py", line 160, in skip
    self.file.seek(n, 1)
  File "/tmp/python-3.7-bin/lib/python3.7/chunk.py", line 113, in seek
    raise RuntimeError
RuntimeError
```

Another second crash comes from sndhdr module itself (again base64 encoded data is first decoded on command line):

```
$ echo AAA= | python3.7 -mbase64 -d > in.file
$ python3.7 sndhdr/test.py in.fileTraceback (most recent call last):
  File "sndhdr/test.py", line 4, in <module>
    sndhdr.what(sys.argv[1])
  File "/tmp/python-3.7-bin/lib/python3.7/sndhdr.py", line 54, in what
    res = whathdr(filename)
  File "/tmp/python-3.7-bin/lib/python3.7/sndhdr.py", line 63, in whathdr
    res = tf(h, f)
  File "/tmp/python-3.7-bin/lib/python3.7/sndhdr.py", line 192, in test_sndr
    rate = get_short_le(h[2:4])
  File "/tmp/python-3.7-bin/lib/python3.7/sndhdr.py", line 213, in get_short_le
    return (b[1] << 8) | b[0]
IndexError: index out of range
```

## What this does
Right now this PR just handles some of the Exceptions caused by bad input and returns `None` based on @vstinner 's [comments in the BPO](https://bugs.python.org/msg321457).

## How did I test this? 
I ran the unit tests and retried the test cases from the bpo post. 
```
(cpython-venv) [ david@lithium ]  ~/Projects/PYTHON/cpython
$>coverage run --pylib Lib/test/test_sndhdr.py
..
----------------------------------------------------------------------
Ran 2 tests in 0.005s

OK
(cpython-venv) [ david@lithium ]  ~/Projects/PYTHON/cpython
$>
```

## Quick questions
So I have two questions about how this should be handled. This is my first python patch so I don't know if there's a better way to do this.

- Should we only handle the specific exceptions that are created in these two examples? or is there a cleaner, more pythonic way to handle it? 

- Should I add test cases for bad inputs to the test_sndhdr.py file? 

Thanks! 

**edit** - sorry about the branch name. typo on my part.

https://bugs.python.org/issue34088

### CODE REVIEW

File: Lib/sndhdr.py [hunk 1]
```diff
@@ -49,10 +49,16 @@
...
 
+
 def what(filename):
     """Guess the type of a sound file."""
-    res = whathdr(filename)
-    return res
+    try:
+        res = whathdr(filename)
+        return res
+    except RuntimeError as e:
```
Comment (ammaraskar) [inline Lib/sndhdr.py]: >Should we only handle the specific exceptions that are created in these two examples? or is there a cleaner, more pythonic way to handle it?

I think this should follow the pattern set out by `sndhdr` and have the exception catching be on a per-test basis. Take a look at this block for example: https://github.com/python/cpython/blob/feabae961707b00008c15a31352e458f4e8b3a6c/Lib/sndhdr.py#L75-L91

You should figure out what exceptions the individual tests code can throw and catch those specifically. As it stands now, its possible for a coding error like an out of bounds exception to be caught accidentally by your exception handler.

**Review by ammaraskar — CHANGES_REQUESTED** [review]

**Comment by ammaraskar** [comment]
>Should I add test cases for bad inputs to the test_sndhdr.py file?

Yes, definitely add test cases.

File: Lib/sndhdr.py (hunk 1 above)
Comment (davidawad) [reply Lib/sndhdr.py]: cool, in that case we already catch exactly those two errors, the `RuntimeError` and the `IndexError` caused by these bad files and catch them explicitly. I'll add some testcases to `Lib/test_sndhdr.py` and @you when it's done.

**Comment by serhiy-storchaka** [comment]
`whathdr()` shouldn't raise exceptions too, except OSErrors when open or read a file. I concur with @ammaraskar, it is better to catch exceptions on a per-test basis.

And please don't make unrelated changes (like adding spaces or newlines).

**Comment by davidawad** [comment]
@ammaraskar just to be sure, are you referring to catching exceptions at the sndhdr.what() level? Or are you asking to add exception handling to the lower level test_<filetype> functions here?

I've added some tests for the different types on inputs this could get, but I want to point out that something weird that I wanted some clarity on. Python hangs when you pass `False` here. 

```python
Python 3.8.0a0 (heads/master:28f07364f0, Jul 17 2018, 11:42:07)
[Clang 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sndhdr
>>> sndhdr.what('NO')
>>> sndhdr.what(None)
>>> sndhdr.what(False)
# process hangs
```
What's the right way to handle that? 

Thanks.

**Comment by ammaraskar** [comment]
>are you referring to catching exceptions at the sndhdr.what() level? Or are you asking to add exception handling to the lower level test_<filetype> functions here?

At the lower test_ level.

>Python hangs when you pass False here. What's the right way to handle this? raise exception for boolean inputs?

Don't handle this case explicitly, what's happening here requires a bit of explanation, even though its not documented open() can apparently open file descriptors by their numbers. False is treated as 0 which opens up stdin for reading, that's why it appears to hang.

Moreover, your tests aren't really testing the primary thing addressed in the bug, which is random binary data. Please add a test that attempts to call `sndhdr.what` on the payloads provided in the bug ticket.

**Comment by davidawad** [comment]
That's really interesting! I figured it was something like that because i'd pass other numbers like `5`, `888`, etc. and I'd get an `OSError` about a bad file descriptor. Good to know! 

I added two different files based on the bug ticket with binary data described in the ticket, and added handlers for the exceptions they throw. 

A few of the builds are still running but let me know if there's anything else i can do. 

Does this warrant an addition to news.d?

File: Lib/sndhdr.py [hunk 2]
```diff
@@ -49,20 +49,28 @@
...
 
+
 def what(filename):
     """Guess the type of a sound file."""
-    res = whathdr(filename)
-    return res
+    try:
+        res = whathdr(filename)
+        return res
+    except (FileNotFoundError) as e:
```
Comment (tiran) [inline Lib/sndhdr.py]: Please remove the extra ``()`` and unused ``as e``.

File: Lib/sndhdr.py [hunk 3]
```diff
@@ -49,20 +49,28 @@
...
+    try:
+        with open(filename, 'rb') as f:
+            h = f.read(512)
+            for tf in tests:
+                res = tf(h, f)
+                if res:
+                    return SndHeaders(*res)
+        return None
+
+    except(TypeError) as e:
```
Comment (tiran) [inline Lib/sndhdr.py]: Same here, ``except TypeError``

File: Lib/sndhdr.py [hunk 4]
```diff
@@ -184,14 +205,19 @@ def test_sndt(h, f):
...
-    if h.startswith(b'\0\0'):
-        rate = get_short_le(h[2:4])
-        if 4000 <= rate <= 25000:
-            return 'sndr', rate, 1, -1, 8
+    try:
+        if h.startswith(b'\0\0'):
+            rate = get_short_le(h[2:4])
+            if 4000 <= rate <= 25000:
+                return 'sndr', rate, 1, -1, 8
+    except (IndexError) as e:
```
Comment (tiran) [inline Lib/sndhdr.py]: Here, too

**Review by tiran — CHANGES_REQUESTED** [review]

File: Lib/sndhdr.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/sndhdr.py]: Why silence FileNotFoundError at first point?

File: Lib/sndhdr.py (hunk 3 above)
Comment (serhiy-storchaka) [inline Lib/sndhdr.py]: I think that no any exception should be ignored at the top level. Catch appropriate exceptions in corresponding test functions.

For example your change silences the error in `sndhdr.what(None)`.

File: Lib/sndhdr.py [hunk 5]
```diff
@@ -72,6 +80,7 @@ def whathdr(filename):
 
 tests = []
 
+
```
Comment (serhiy-storchaka) [inline Lib/sndhdr.py]: Unrelated change.

File: Lib/sndhdr.py [hunk 6]
```diff
@@ -158,13 +171,20 @@ def test_wav(h, f):
     # 'RIFF' <len> 'WAVE' 'fmt ' <len>
     if not h.startswith(b'RIFF') or h[8:12] != b'WAVE' or h[12:16] != b'fmt ':
         return None
-    f.seek(0)
     try:
+        f.seek(0)
         w = wave.open(f, 'r')
-    except (EOFError, wave.Error):
+    except (EOFError,
```
Comment (serhiy-storchaka) [inline Lib/sndhdr.py]: Can't it be written in a single line?

File: Lib/sndhdr.py (hunk 5 above)
Comment (davidawad) [reply Lib/sndhdr.py]: This was a pep8 recommendation from a linting tool that I use. I'll remove it.

File: Lib/sndhdr.py (hunk 4 above)
Comment (davidawad) [reply Lib/sndhdr.py]: Will do.

File: Lib/sndhdr.py (hunk 6 above)
Comment (davidawad) [reply Lib/sndhdr.py]: In this case it's just the 4 error types, if you think it makes more sense to put them in one line I can definitely move them. 

Just wondering why? Is it more _pythonic_ to do it that way?

File: Lib/sndhdr.py (hunk 3 above)
Comment (davidawad) [reply Lib/sndhdr.py]: yes, definitely. My catch in `sndhdr.what` should really be happening here. I'll move it. Thanks!

Comment (davidawad) [reply Lib/sndhdr.py]: @tiran will do. definitely was silly to create `e` and not use it. thanks

File: Lib/sndhdr.py (hunk 2 above)
Comment (davidawad) [reply Lib/sndhdr.py]: @serhiy-storchaka This was my mistake. This exception really happens inside of `whathdr()`. I'll catch it there.

Comment (davidawad) [reply Lib/sndhdr.py]: @serhiy-storchaka I also added unit tests specifically checking for bad input on both `what()` and `whathdr()` 

I should have caught this bug when writing those tests. My bad on that, still trying to learn and get better here :)

**Comment by davidawad** [comment]
I have made the requested changes; please review again. 

^^ comment for bot

File: Lib/sndhdr.py [hunk 7]
```diff
@@ -57,12 +58,17 @@ def what(filename):
...
+    try:
+        with open(filename, 'rb') as f:
+            h = f.read(512)
+            for tf in tests:
+                res = tf(h, f)
+                if res:
+                    return SndHeaders(*res)
+        return None
+
+    except (TypeError,
```
Comment (serhiy-storchaka) [inline Lib/sndhdr.py]: Don't catch TypeError here. Catch it in test functions.

File: Lib/sndhdr.py (hunk 6 above)
Comment (serhiy-storchaka) [reply Lib/sndhdr.py]: Because
it
is
better
to
not
split
a
line
without
a
need.

File: Lib/sndhdr.py [hunk 8]
```diff
@@ -90,6 +96,7 @@ def test_aifc(h, f):
     return (fmt, a.getframerate(), a.getnchannels(),
             a.getnframes(), 8 * a.getsampwidth())
 
+
```
Comment (serhiy-storchaka) [inline Lib/sndhdr.py]: This is not related change.

File: Lib/sndhdr.py [hunk 9]
```diff
@@ -57,12 +58,17 @@ def what(filename):
...
+        with open(filename, 'rb') as f:
+            h = f.read(512)
+            for tf in tests:
+                res = tf(h, f)
+                if res:
+                    return SndHeaders(*res)
+        return None
+
+    except (TypeError,
+            FileNotFoundError):
```
Comment (serhiy-storchaka) [inline Lib/sndhdr.py]: Don't catch FileNotFoundError.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
Please remove all unrelated changes and catch exceptions in proper places.

File: Lib/sndhdr.py (hunk 8 above)
Comment (davidawad) [reply Lib/sndhdr.py]: consider it gone!

File: Lib/sndhdr.py (hunk 7 above)
Comment (davidawad) [reply Lib/sndhdr.py]: So this TypeError is thrown when passing `None` to the `open()` call. That's specific to this function, would you prefer to just not catch it?  

let me know whatever you'd prefer and i'll make the change.


---


When commenting out the try catch this is what happens in `whathdr`


```
>>> sndhdr.whathdr(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/david/Projects/PYTHON/cpython/Lib/sndhdr.py", line 62, in whathdr
    with open(filename, 'rb') as f:
TypeError: expected str, bytes or os.PathLike object, not NoneType
>>>
```

File: Lib/sndhdr.py (hunk 9 above)
Comment (davidawad) [reply Lib/sndhdr.py]: done

File: Lib/sndhdr.py (hunk 7 above)
Comment (ammaraskar) [reply Lib/sndhdr.py]: I don't think you should catch it, from the traceback its very obvious what went wrong, since the documentation explicitly says `what()` works on files, it is expected to receive FileNotFound and invalid file name errors.

File: Lib/test/test_sndhdr.py [hunk 10]
```diff
@@ -25,6 +26,30 @@ def test_data(self):
             self.assertEqual(what.nframes, expected[3])
             self.assertEqual(what.sampwidth, expected[4])
 
+    def test_badinputs(self):
```
Comment (ammaraskar) [inline Lib/test/test_sndhdr.py:28]: Since all of these are expected to fail and return `None`, I'd suggest simplifying those loops to:

```python
for filename in ['input1.bad', 'input2.bad']:
    ...
    self.assertEqual(whathdr, None)
```

File: Lib/sndhdr.py (hunk 7 above)
Comment (davidawad) [reply Lib/sndhdr.py]: wholly agree. I'll make the change.

File: Lib/test/test_sndhdr.py (hunk 10 above)
Comment (davidawad) [reply Lib/test/test_sndhdr.py:28]: that's fine by me! I'll make the change

File: Lib/sndhdr.py [hunk 11]
```diff
@@ -106,7 +107,7 @@ def test_au(h, f):
     encoding = func(h[12:16])
     rate = func(h[16:20])
     nchannels = func(h[20:24])
-    sample_size = 1 # default
+    sample_size = 1  # default
```
Comment (vstinner) [inline Lib/sndhdr.py]: Please revert coding style only changes.

File: Lib/sndhdr.py [hunk 12]
```diff
@@ -158,13 +162,17 @@ def test_wav(h, f):
     # 'RIFF' <len> 'WAVE' 'fmt ' <len>
     if not h.startswith(b'RIFF') or h[8:12] != b'WAVE' or h[12:16] != b'fmt ':
         return None
-    f.seek(0)
     try:
+        f.seek(0)
         w = wave.open(f, 'r')
-    except (EOFError, wave.Error):
+    except (EOFError, RuntimeError, AttributeError, wave.Error):
```
Comment (vstinner) [inline Lib/sndhdr.py]: Catching AttributeError or RuntimeError *here* is wrong. wave.open() should be modified to better handle errors.

File: Lib/sndhdr.py [hunk 13]
```diff
@@ -158,13 +162,17 @@ def test_wav(h, f):
...
         return None
-    return ('wav', w.getframerate(), w.getnchannels(),
-                   w.getnframes(), 8*w.getsampwidth())
+    return ('wav',
+            w.getframerate(),
+            w.getnchannels(),
+            w.getnframes(),
+            8 * w.getsampwidth())
```
Comment (vstinner) [inline Lib/sndhdr.py]: Ditto: revert coding style only changes.

File: Lib/sndhdr.py [hunk 14]
```diff
@@ -184,14 +193,19 @@ def test_sndt(h, f):
...
-    if h.startswith(b'\0\0'):
-        rate = get_short_le(h[2:4])
-        if 4000 <= rate <= 25000:
-            return 'sndr', rate, 1, -1, 8
+    try:
+        if h.startswith(b'\0\0'):
+            rate = get_short_le(h[2:4])
+            if 4000 <= rate <= 25000:
+                return 'sndr', rate, 1, -1, 8
+    except IndexError:
```
Comment (vstinner) [inline Lib/sndhdr.py]: What is the IndexError? Is it when h is empty?

**Review by vstinner — CHANGES_REQUESTED** [review]

File: Lib/sndhdr.py (hunk 11 above)
Comment (davidawad) [reply Lib/sndhdr.py]: done!

File: Lib/sndhdr.py (hunk 14 above)
Comment (davidawad) [reply Lib/sndhdr.py]: Yeah. 

If you remove the try catch and call the function on the second bad input file, it tries to index it when it's really empty. 
```
>>> sndhdr.what('Lib/test/sndhdrdata/input2.bad')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/david/Projects/PYTHON/cpython/Lib/sndhdr.py", line 54, in what
    res = whathdr(filename)
  File "/Users/david/Projects/PYTHON/cpython/Lib/sndhdr.py", line 63, in whathdr
    res = tf(h, f)
  File "/Users/david/Projects/PYTHON/cpython/Lib/sndhdr.py", line 193, in test_sndr
    rate = get_short_le(h[2:4])
  File "/Users/david/Projects/PYTHON/cpython/Lib/sndhdr.py", line 218, in get_short_le
    return (b[1] << 8) | b[0]
IndexError: index out of range
>>>
```


Looking more specifically at it, `input2.bad` has a basically empty header. 

```python
>>> with open ('Lib/test/sndhdrdata/input2.bad') as f:
...     h = f.read(512)
>>> h
'\x00\x00'
# h[2:4] gets passed to get_short_le
>>> h[2:4]
''
```
because `h[2:4]` is empty when it gets passed to `get_short_le(b)`, `b[1]` fails. 

@vstinner  I think it would be consistent with the feedback I've gotten so far to move the exception handler to `get_short_le` instead of it's calling function, I can make that change, but I'd want to know what you think first.

File: Lib/sndhdr.py (hunk 12 above)
Comment (davidawad) [reply Lib/sndhdr.py]: @vstinner I think you're probably right. Here's what I was able to figure out. 

If I remove the `RuntimeError` from the catch list, this is what happens in the python run. 

```python
david@lithium ~/P/P/cpython>pso-34088>./python.exe                              10:40:56
Python 3.8.0a0 (heads/master:28f07364f0, Jul 17 2018, 11:42:07)
[Clang 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sndhdr
>>> sndhdr.what('./Lib/test/sndhdrdata/input1.bad')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/david/Projects/PYTHON/cpython/Lib/sndhdr.py", line 54, in what
    res = whathdr(filename)
  File "/Users/david/Projects/PYTHON/cpython/Lib/sndhdr.py", line 63, in whathdr
    res = tf(h, f)
  File "/Users/david/Projects/PYTHON/cpython/Lib/sndhdr.py", line 162, in test_wav
    w = wave.open(f, 'r')
  File "/Users/david/Projects/PYTHON/cpython/Lib/wave.py", line 510, in open
    return Wave_read(f)
  File "/Users/david/Projects/PYTHON/cpython/Lib/wave.py", line 164, in __init__
    self.initfp(f)
  File "/Users/david/Projects/PYTHON/cpython/Lib/wave.py", line 153, in initfp
    chunk.skip()
  File "/Users/david/Projects/PYTHON/cpython/Lib/chunk.py", line 160, in skip
    self.file.seek(n, 1)
  File "/Users/david/Projects/PYTHON/cpython/Lib/chunk.py", line 113, in seek
    raise RuntimeError
RuntimeError
>>>
```


--- 


I don't have a complete understand what's happening inside `chunk.py`, it looks like seek is being asked to go too far in the file and it's throwing a runtime error.

I added a quick print inside `chunk.py` (line 112) to `self.seek()` and this is what came out of my test. 
  
#### my change 
```python 
    def seek(self, pos, whence=0):
        """Seek to specified position into the chunk.
        Default position is 0 (start of chunk).
        If the file is not seekable, this will result in an error.
        """

        if self.closed:
            raise ValueError("I/O operation on closed file")
        if not self.seekable:
            raise OSError("cannot seek")
        if whence == 1:
            pos = pos + self.size_read
        elif whence == 2:
            pos = pos + self.chunksize
        if pos < 0 or pos > self.chunksize:
            print('SEEKING TO POSITION: ', pos, "CHUNK SIZE:", self.chunksize )
            raise RuntimeError
        self.file.seek(self.offset + pos, 0)
        self.size_read = pos
```


#### output of that test
```python
Python 3.8.0a0 (heads/master:28f07364f0, Jul 17 2018, 11:42:07)
[Clang 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sndhdr
>>> sndhdr.what('Lib/test/sndhdrdata/input1.bad')
SEEKING TO POSITION:  808464444 CHUNK SIZE: 808464432  ## pos > self.chunksize
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/david/Projects/PYTHON/cpython/Lib/sndhdr.py", line 54, in what
    res = whathdr(filename)
# . . .  
```



given that `skip()` passes on `OSError`, maybe it makes sense to ignore the `RuntimeError` from `seek()`? 



As for the `AttributeError` I'm not able to reproduce it when running my tests again with these files which is _strange_. I'll remove the catch for `AttributeError`, sorry about that.

Comment (vstinner) [reply Lib/sndhdr.py]: You are free to decide how to modify the code, but IMHO wave constructor should be modified to raise a wave.Error if the file is invalid.

File: Lib/sndhdr.py (hunk 14 above)
Comment (vstinner) [reply Lib/sndhdr.py]: in that case, check len(h) instead of *expecting* an IndexError.

Comment (davidawad) [reply Lib/sndhdr.py]: lol this is a good point. I'll check the length. thanks!

File: Lib/sndhdr.py (hunk 12 above)
Comment (davidawad) [reply Lib/sndhdr.py]: Sounds good, raising a wave.error definitely makes more sense than a RuntimeError here.

File: Lib/sndhdr.py (hunk 14 above)
Comment (davidawad) [reply Lib/sndhdr.py]: If the `len(h) < 2`, what should `get_short_le()` do? just return `None`?
the calling method compares it a bit,  `if 4000 <= rate <= 25000`. 
given that these subroutines are supposed to extract numbers from bytes, I'm going to return a `0` on these unless anyone has any objections. 

my question is, should that bounds check also apply to the other `get_x_x()` subroutines?

Comment (serhiy-storchaka) [reply Lib/sndhdr.py]: Catching an IndexError and checking `len(h)` look equally good to me.

File: Lib/sndhdr.py [hunk 15]
```diff
@@ -158,10 +157,10 @@ def test_wav(h, f):
     # 'RIFF' <len> 'WAVE' 'fmt ' <len>
     if not h.startswith(b'RIFF') or h[8:12] != b'WAVE' or h[12:16] != b'fmt ':
         return None
-    f.seek(0)
     try:
+        f.seek(0)
```
Comment (serhiy-storchaka) [inline Lib/sndhdr.py]: Why this line has been moved?

File: Lib/sndhdr.py [hunk 16]
```diff
@@ -175,23 +174,27 @@ def test_8svx(h, f):
     # Should decode it to get #channels -- assume always 1
     return '8svx', 0, 1, 0, 8
 
-tests.append(test_8svx)
 
+tests.append(test_8svx)
```
Comment (serhiy-storchaka) [inline Lib/sndhdr.py]: Why this line has been moved?

File: Lib/sndhdr.py (hunk 13 above)
Comment (davidawad) [reply Lib/sndhdr.py]: reverted.

File: Lib/sndhdr.py (hunk 14 above)
Comment (davidawad) [reply Lib/sndhdr.py]: I'll go for checking `len()`.

File: Lib/sndhdr.py (hunk 15 above)
Comment (davidawad) [reply Lib/sndhdr.py]: This didn't need to be moved, that was my mistake!

File: Lib/sndhdr.py (hunk 16 above)
Comment (davidawad) [reply Lib/sndhdr.py]: I had some weirdness with the newlines, I've fixed it.

**Comment by davidawad** [comment]
Alright, this was a really good learning experience. Thank you all for sticking with me and helping me make this change the right way.
 
Would you prefer a rebase to compress all these commits into one since there's a lot of commits here? 

and for the bot: 
I have made the requested changes; please review again

**Comment by davidawad** [comment]
Can anyone review this? It's been a while and I'd love to see it go in.

File: Lib/test/test_sndhdr.py [hunk 17]
```diff
@@ -25,6 +25,24 @@ def test_data(self):
...
+
+        # test feeding bad binary data to sound header functions
+        for filename in ['input1.bad', 'input2.bad']:
+
+            filename = findfile(filename, subdir="sndhdrdata")
+
+            what = sndhdr.what(filename)
+            self.assertEqual(what, None)
```
Comment (serhiy-storchaka) [inline Lib/test/test_sndhdr.py]:
```suggestion
            self.assertIsNone(what)
```

File: Lib/test/test_sndhdr.py [hunk 18]
```diff
@@ -25,6 +25,24 @@ def test_data(self):
...
+
+            whathdr = sndhdr.whathdr(filename)
+            self.assertEqual(whathdr, None)
```
Comment (serhiy-storchaka) [inline Lib/test/test_sndhdr.py]:
```suggestion
            self.assertIsNone(whathdr)
```

File: Lib/test/test_sndhdr.py [hunk 19]
```diff
@@ -25,6 +25,24 @@ def test_data(self):
...
+
+            for tf in sndhdr.tests:
+                with open(filename, 'rb') as f:
+                    h = f.read(512)
+                    self.assertEqual(tf(h, f), None)
```
Comment (serhiy-storchaka) [inline Lib/test/test_sndhdr.py]:
```suggestion
                    self.assertIsNone(tf(h, f))
```

File: Lib/wave.py [hunk 20]
```diff
@@ -150,7 +150,11 @@ def initfp(self, file):
                 self._nframes = chunk.chunksize // self._framesize
                 self._data_seek_needed = 0
                 break
-            chunk.skip()
+            try:
+                chunk.skip()
+            except RuntimeError:
```
Comment (serhiy-storchaka) [inline Lib/wave.py]: Why this was added? Tests are passed without this change.

**Comment by davidair** [comment]
> @tiran @serhiy-storchaka @vstinner Re-requesting review to hopefully push this one to be merged.
> 
> @davidair, you don't have to squash the commits as that is done upon merge. Also, we don't generally accept PRs that are just PEP8 changes as they create more noise in the git blame than benefit. Thank you for asking and wanting to improve Python! 🙂

Probably wrong David - did you mean @davidawad?

**Comment by csabella** [comment]
> Probably wrong David - did you mean @davidawad?

Yes, sorry about that and thank you for including the correct David.

File: Lib/wave.py (hunk 20 above)
Comment (davidawad) [reply Lib/wave.py]: I don't think they actually pass without catching this. It's been a while so looking at this again I get the following issue when running these tests without this fix. 

```
david@lithium ~/P/P/c/L/test>pso-34088>../../python.exe test_sndhdr.py          18:22:08
E..
======================================================================
ERROR: test_badinputs (__main__.TestFormats)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_sndhdr.py", line 35, in test_badinputs
    what = sndhdr.what(filename)
  File "/Users/david/Projects/PYTHON/cpython/Lib/sndhdr.py", line 54, in what
    res = whathdr(filename)
  File "/Users/david/Projects/PYTHON/cpython/Lib/sndhdr.py", line 63, in whathdr
    res = tf(h, f)
  File "/Users/david/Projects/PYTHON/cpython/Lib/sndhdr.py", line 163, in test_wav
    w = wave.open(f, 'r')
  File "/Users/david/Projects/PYTHON/cpython/Lib/wave.py", line 510, in open
    return Wave_read(f)
  File "/Users/david/Projects/PYTHON/cpython/Lib/wave.py", line 164, in __init__
    self.initfp(f)
  File "/Users/david/Projects/PYTHON/cpython/Lib/wave.py", line 153, in initfp
    chunk.skip()
  File "/Users/david/Projects/PYTHON/cpython/Lib/chunk.py", line 160, in skip
    self.file.seek(n, 1)
  File "/Users/david/Projects/PYTHON/cpython/Lib/chunk.py", line 113, in seek
    raise RuntimeError
RuntimeError

----------------------------------------------------------------------
Ran 3 tests in 0.010s

FAILED (errors=1)
```

Comment (davidawad) [reply Lib/wave.py]: I believe at the time I thought that this file should have reported the runtime error if it failed to skip through a wav file that was poorly formatted. 

Given I'm not very experienced at being a principled programmer; @serhiy-storchaka where do you think is the best place to catch this error or deal with the tests failing on the bad input files?

**Comment by davidawad** [comment]
Definitely, thanks for bringing this back to my attention @csabella . I really want to see this somewhat simple PR get merged as it's been a really good learning experience overall and i'd love to try to improve upon this and help out on some of the other easier python issues as well. 

@serhiy-storchaka If I didn't make it clear last year I really appreciate your patience and advice it's been really helpful.

File: Lib/test/test_sndhdr.py (hunk 17 above)
Comment (davidawad) [reply Lib/test/test_sndhdr.py]: done

File: Lib/test/test_sndhdr.py (hunk 18 above)
Comment (davidawad) [reply Lib/test/test_sndhdr.py]: done

File: Lib/test/test_sndhdr.py (hunk 19 above)
Comment (davidawad) [reply Lib/test/test_sndhdr.py]: done

**Comment by serhiy-storchaka** [comment]
Both `input1.bad` and `input2.bad` are empty files.

Since all errors are due to truncation, I suggest you to use the following test:
```python
    def test_truncated(self):
        for filename, expected in SAMPLES:
            filename = findfile(filename, subdir="sndhdrdata")
            with open(filename, 'rb') as f:
                data = f.read()
            self.addCleanup(unlink, TESTFN)
            with self.subTest(filename=filename):
                for i in range(len(data)):
                    with open(TESTFN, 'wb') as f:
                        f.write(data[:i])
                    what = sndhdr.what(TESTFN)
                    if what is not None:
                        self.assertEqual(what, expected)
```
`SAMPLES` is the same as in `test_data`, `TESTFN` and `unlink` are imported from `test.support`.

Currently this test fails. Make it passing.

Maybe the simpler test is enough: `self.assertEqual(what[0], expected[0])` instead of `self.assertEqual(what, expected)`, but it would be better to achieve the success of the more strong test.

**Comment by davidawad** [comment]
I have imported that test and i'm able to correct some of these failures but i don't really understand exactly how these audio formats work. Is there documentation I can read to get a better understanding of how to handle some of the failure cases this test causes?

**Comment by serhiy-storchaka** [comment]
In general you need to catch all expected errors in `test_*` functions and return None. You have to remove `if len(b) < 4:` from `get_long_be()` and catch an IndexError in functions that use it.

**Comment by serhiy-storchaka** [comment]
Alternatively you can add `if len(h) < 24: return None` in `test_au()` and similar checks in other functions. Then you could use `int.from_bytes()` instead of `get_long_be()` and friends.

**Comment by taleinat** [comment]
Ping, @davidawad!

**Comment by taleinat** [comment]
Ping, @davidawad!

**Comment by davidawad** [comment]
@taleinat admittedly I gave up on this after a while. I'll try to take another look this weekend and see if I can figure it out. 

It was tricky just because I didn't have anyone to really ask questions to but I'm sure we can get it done.

**Comment by taleinat** [comment]
@davidawad, please feel free to ask any questions! You can do so here, on the b.p.o. issue, on the Python IRC or Zulip channels, or on the various mailing lists.

**Comment by hugovk** [comment]
Note the sndhdr module is deprecated in 3.11 and set for removal in 3.13.

See [PEP 594 – Removing dead batteries from the standard library](https://peps.python.org/pep-0594/) and https://github.com/python/cpython/issues/91217.

**Comment by kumaraditya303** [comment]
Closing as the module is deprecated.
